std/experimental/typecons.d

```diff
@@ -1074,3 +1074,258 @@ pure nothrow @system unittest
         arr.ptr++
     ));
 }
+
+import std.traits : isInstanceOf;
+
+/** Wrapper type for an unexpected value of type `E`.
+ */
+private struct Unexpected(E)
+{
+    E value;
+    alias value this;
+}
+
+/** Union (sum) type of either an expected (most probable) value of type `T` or
+ * an unexpected value of type `E` (being an instance of type `Unexpected!E`).
+ *
+ * `E` is typically an error code (for instance C/C++'s' `errno` int) or a
+ * subclass of `Exception` (which is the default).
+ *
+ * See_Also: https://www.youtube.com/watch?v=nVzgkepAg5Y
+ * See_Also: https://github.com/dlang/phobos/pull/6665
+ *
+ * TODO https://dlang.org/phobos/std_typecons.html#.apply
+ *
+ * TODO I'm not convinced about the naming
+ * - `Expected`: instead call it something that tells us that it can be either expected or unexpected?
+ * - `Unexpected`: if so why shouldn't we have a similar value wrapper `Expected`?
+ *
+ * TODO we could get around the `Unexpected` wrapper logic by instead expressing
+ * construction in static constructor functions, say,:
+ * - static typeof(this) fromExpectedValue(T expectedValue)
+ * - static typeof(this) fromUnexpectedValue(E unexpectedValue)
+ *
+ * TODO swap
+ *
+ * TODO which functions should be `nothrow`?
+ *
+ * TODO later on: remove _ok when `_expectedValue` and ` _unexpectedValue` can store this state
+ * "collectively" for instance when both are pointers or classes (use trait
+ * `isAddress`)
+ *
+ * TODO ok to default `E` to `Exception`?
+ */
+struct Expected(T, E = Exception)
+if (!isInstanceOf!(Unexpected, T)) // an `Unexpected` cannot be `Expected` :)
+{
+    // TODO ok for default construction to initialize
+    // - _expectedValue = T.init (zeros)
+    // - _ok = true (better to have _isError so default is zero bits here aswell?)
+
+    /// Construct from expected value `expectedValue.`
+    this(T expectedValue) @trusted
+    {
+        // TODO reuse opAssign?
+        _expectedValue = expectedValue;       // TODO use `moveEmplace` here aswell?
```

Comment:
I'd probably use `moveEmplace`.
Because it's part of an union, you must be careful and default initialize `_expectedValue` with `T.init`.

Take the following scenario:
If `T` has a `ptr` and `opAssign` does some checks on `ptr` (like for reference counting) then you have UB.

Reply:
Done.